instrumentation/kotlin-coroutines/javaagent/src/test/kotlin/KotlinCoroutineTests.kt

```diff
@@ -145,35 +147,35 @@ class KotlinCoroutineTests(private val dispatcher: CoroutineDispatcher) {
   suspend fun a(iter: Long) {
     var span = tracer.spanBuilder("a").startSpan()
     span.setAttribute("iter", iter)
-    var scope = span.makeCurrent()
-    delay(10)
-    a2(iter)
-    scope.close()
+    withContext(span.asContextElement()) {
```

Comment:
These changes are the key to the proposal. My theory is `span.makeCurrent` is illegal code - the only reason it works here is because the agent is present, but if the agent was not present, resumed coroutines could see completely unrelated contexts, which is a disaster. The idea is that our agent is here to enable tracing, not to allow illegal code. How to "codify" that it's illegal is a tricky point though but something we will want to think about. Current idea is that https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/testing/src/main/java/io/opentelemetry/sdk/testing/context/StrictContextStorage.java#L99 can also check the stacktrace to see if it is being called from a coroutine, and not from the context element. Dunno if it'll work but if it does, I think we end up in a nice place :)

Reply:
Also kudos to @adriancole for giving the idea of stack-walking as a potential mechanism for checking

Reply:
Does this require application authors who use auto-instrumentation and simply use (e.g.) kotlin suspend functions in their logic (without any manual instrumentation) to explicitly use (kotlin) `withContext` to pass our (otel) context?  That is, if app logic is written in kotlin and taking advantage of auto-instrumentation of inbound and outbound connections, will this still work without the app author having to take extra steps?

Reply:
Don't think so - the agent with this code is making sure to automatically propagate the parent at the time the coroutine is launched into its context, so that parent is always available. If all they're doing is using libraries which are auto-instrumented by the agent, it should work fine without calling `withContext`. It's only when doing manual instrumentation within Kotlin that they need to use `withContext` instead of `makeCurrent`. How does that sound?

Reply:
I can see the confusion of this particular test, since it's a javaagent test, but also uses the library instrumentation directly.

do you think we need the nested spans to validate `launchConcurrentSuspendFunctions`? if not, we could remove `a2` and `b2` and not create any nested scopes at all, and not need to use the library instrumentation in the test.

if you think the nested scopes is a good test showing interop, maybe we can move it out into a separate test class `...InteropTest`.

Reply:
The nesting spans are crucial to the test - without any form of instrumentation, we were previously reporting nonsensical traces like "a causes b causes a2" and "b2 causes a".

Reply:
> . It's only when doing manual instrumentation within Kotlin that they need to use `withContext` instead of `makeCurrent`. How does that sound?

That sounds great; I have no problem making kotlin instrumentation authors work a little to preserve (otel) context.  I actually originally tried something like this (using the kotlin coroutine context system more "naturally) but was unable to make it work.  I'm glad you were able to figure it out!  l'm going to try to find some time to manually validate with a "real" kotlin app and see how this change works out in practice.

Reply:
@johnbley #1620 links to a real, though simple kotlin app that exhibits problems with span propagation